[EV-3369] Add management and managed cluster installation to Helm Install (#881)
tigera/docs: calico-enterprise/getting-started/install-on-clusters/kubernetes/helm.mdx

If you are installing on a cluster installed by EKS, GKE, AKS or Mirantis Kubernetes Engine (MKE), or you need to customize TLS certificates, you **must** customize this Helm chart by creating a `values.yaml` file. Otherwise, you can skip this step.

1. If you are installing on a cluster installed by EKS, GKE, AKS or Mirantis Kubernetes Engine (MKE), set the `kubernetesProvider` as described in the [Installation reference](../../../reference/installation/api.mdx#operator.tigera.io/v1.Provider). For example:

  ```bash
  echo '{ installation: {kubernetesProvider: EKS }}' > values.yaml
  ```

   For Azure AKS cluster with no Kubernetes CNI pre-installed, create `values.yaml` with the following command:

  ```yaml
  cat > values.yaml <<EOF
  installation:
    kubernetesProvider: AKS
    cni:
      type: Calico
    calicoNetwork:
      bgp: Disabled
      ipPools:
      - cidr: 10.244.0.0/16
        encapsulation: VXLAN
  EOF
  ```

1. Add any other customizations you require to `values.yaml`. To see values that can be customized in the chart, see the [helm docs](https://helm.sh/docs/) or run the following command:

  ```bash
  helm show values tigera-operator-{{version}}.tgz
  ```

>    to see the values that can be customized in the chart.

Reviewer: ```suggestion
delete this line...straggler from edit.
```

### Install {{prodname}}

<Tabs>

<TabItem label="Standalone" value="Standalone">

> 1. [Configure a storage class for Calico Enterprise](../../../operations/logstorage/create-storage)

Reviewer: ```suggestion
1. [Configure a storage class for Calico Enterprise](../../../operations/logstorage/create-storage).
```

  ```bash
  helm install calico-enterprise tigera-operator-{{version}}.tgz \
  --set-file imagePullSecrets.tigera-pull-secret=<path/to/pull/secret>,tigera-prometheus-operator.imagePullSecrets.tigera-pull-secret=<path/to/pull/secret> \
  --set licenseKey.enabled=true \
  --set licenseKey.certificate=$LICENSE_KEY_CERTIFICATE \
  --set licenseKey.token=$LICENSE_KEY_TOKEN \
  --namespace tigera-operator --create-namespace
  ```

   or if you created a `values.yaml` above:

  ```bash
  helm install calico-enterprise tigera-operator-{{version}}.tgz -f values.yaml \
  --set-file imagePullSecrets.tigera-pull-secret=<path/to/pull/secret>,tigera-prometheus-operator.imagePullSecrets.tigera-pull-secret=<path/to/pull/secret> \
  --set licenseKey.enabled=true \
  --set licenseKey.certificate=$LICENSE_KEY_CERTIFICATE \
  --set licenseKey.token=$LICENSE_KEY_TOKEN \
  --namespace tigera-operator --create-namespace
  ```

1. You can now monitor progress with the following command:

  ```bash
  watch kubectl get tigerastatus
  ```

  Congratulations! You have now installed {{prodname}} using the Helm 3 chart.

</TabItem>
<TabItem label="Management" value="Management">

> 1. [Configure a storage class for Calico Enterprise](../../../operations/logstorage/create-storage)

Reviewer: ```suggestion
1. [Configure a storage class for Calico Enterprise](../../../operations/logstorage/create-storage).
```

1. Export the service port number, and the public IP or host of the management cluster. (Ex. "example.com:1234" or "10.0.0.10:1234".)

  ```bash
  export MANAGEMENT_CLUSTER_ADDR=<your-management-cluster-addr>
  ```

1. Export the service type (`NodePort` or `LoadBalancer`), the service port, target port and protocol.
  
  ```bash
  export EXT_SERVICE_TYPE=<external-service-type>
  export EXT_SERVICE_PORT=<external-service-port>
  export EXT_SERVICE_TARGET_PORT=<external-service-target-port>
  export EXT_SERVICE_PROTOCOL=<external-service-protocol>
  ```
  
  If the external service is of type NodePort.

  ```bash
  export EXT_SERVICE_TYPE=<external-service-node-port>
  ```

> 1. Export 1..n managed clusters.

Reviewer: I've not seen it this way before, but if more common for developers, ok. Or, 1-n.

Author: I've changed this to `one or more`

  ```bash
  helm install {{prodnamedash}} tigera-operator-{{version}} -f values.yaml \
  --set-file imagePullSecrets.tigera-pull-secret=<path/to/pull/secret>,tigera-prometheus-operator.imagePullSecrets.tigera-pull-secret=<path/to/pull/secret> \
  
  --set managementCluster.enabled=true \
  --set managementCluster.address=$MANAGEMENT_CLUSTER_ADDR \
  --set managementCluster.service.type=$EXT_SERVICE_TYPE \
  --set managementCluster.service.port=$EXT_SERVICE_PORT \
  --set managementCluster.service.targetPort=$EXT_SERVICE_TARGET_PORT \
  --set managementCluster.service.protocol=$EXT_SERVICE_PROTOCOL \
  --set managementCluster.service.nodePort=$EXT_SERVICE_NODE_PORT \
  
  --set managedClusters.enabled=true \
  --set managedClusters.clusters[0].name=$MANAGED_CLUSTER_NAME_1 \
  --set managedClusters.clusters[0].operatorNamespace=$MANAGED_CLUSTER_OPERATOR_NAMESPACE \
  --set managedClusters.clusters[0].certificate=$MANAGED_CLUSTER_CERTIFICATE_1 \

  --set licenseKey.enabled=true \
  --set licenseKey.certificate=$LICENSE_KEY_CERTIFICATE \
  --set licenseKey.token=$LICENSE_KEY_TOKEN \

  --namespace tigera-operator --create-namespace
  ```

>   or with a LoadBalancer on AWS add expected annotations to the above installation.

Reviewer: ```suggestion
  or with a LoadBalancer on AWS, add expected annotations to the above installation.
```

  ```bash
  --set managementCluster.service.annotations[0].key="service.beta.kubernetes.io/aws-load-balancer-type" \
  --set managementCluster.service.annotations[0].value=external \
  --set managementCluster.service.annotations[1].key="service.beta.kubernetes.io/aws-load-balancer-nlb-target-type" \
  --set managementCluster.service.annotations[1].value=instance \
  --set managementCluster.service.annotations[2].key="service.beta.kubernetes.io/aws-load-balancer-scheme" \
  --set managementCluster.service.annotations[2].value=internet-facing \
  ```

1. You can now monitor progress with the following command:

  ```bash
  watch kubectl get tigerastatus
  ```

  Congratulations! You have now installed {{prodname}} using the Helm 3 chart.

</TabItem>
<TabItem label="Managed" value="Managed">

> 1. [Configure a storage class for Calico Enterprise](../../../operations/logstorage/create-storage)

Reviewer: ```suggestion
1. [Configure a storage class for Calico Enterprise](../../../operations/logstorage/create-storage).
```

  ```

### Customize the Helm chart

If you are installing on a cluster installed by EKS, GKE, AKS or Mirantis Kubernetes Engine (MKE), or you need to customize TLS certificates, you **must** customize this Helm chart by creating a `values.yaml` file. Otherwise, you can skip this step.

1. If you are installing on a cluster installed by EKS, GKE, AKS or Mirantis Kubernetes Engine (MKE), set the `kubernetesProvider` as described in the [Installation reference](../../../reference/installation/api.mdx#operator.tigera.io/v1.Provider). For example:

  ```bash
  echo '{ installation: {kubernetesProvider: EKS }}' > values.yaml
  ```

   For Azure AKS cluster with no Kubernetes CNI pre-installed, create `values.yaml` with the following command:

  ```yaml
  cat > values.yaml <<EOF
  installation:
    kubernetesProvider: AKS
    cni:
      type: Calico
    calicoNetwork:
      bgp: Disabled
      ipPools:
      - cidr: 10.244.0.0/16
        encapsulation: VXLAN
  EOF
  ```

1. Add any other customizations you require to `values.yaml`. To see values that can be customized in the chart, see the [helm docs](https://helm.sh/docs/) or run the following command:

  ```

>   helm show values projectcalico/tigera-operator --version {{releaseTitle}}

Reviewer 2: For master, the version must be v0.0. So when this doc is ready, make sure the master version of this page is slightly different.

Author: Thanks, there is now a if-else addressing the `master` version case

1. [Configure a storage class for Calico Enterprise](../../../operations/logstorage/create-storage).

1. Export the service port number, and the public IP or host of the management cluster. (Ex. "example.com:1234" or "10.0.0.10:1234".)

  ```bash
  export MANAGEMENT_CLUSTER_ADDR=<your-management-cluster-addr>
  ```

1. Export the service type (`NodePort` or `LoadBalancer`), the service port, target port and protocol.
  
  ```bash
  export EXT_SERVICE_TYPE=<external-service-type>
  export EXT_SERVICE_PORT=<external-service-port>
  export EXT_SERVICE_TARGET_PORT=<external-service-target-port>
  export EXT_SERVICE_PROTOCOL=<external-service-protocol>
  ```
  
  If the external service is of type NodePort.

  ```bash
  export EXT_SERVICE_TYPE=<external-service-node-port>
  ```

1. Export one or more managed clusters.

  ```bash
  export MANAGED_CLUSTER_NAME=<managed-cluster-name>
  export MANAGED_CLUSTER_OPERATOR_NAMESPACE=<managed-cluster-operator-namespace>
  export MANAGED_CLUSTER_CERTIFICATE=<managed-cluster-certificate>
  ```

1. Install the Tigera {{prodname}} operator and custom resource definitions using the Helm 3 chart, with a NodePort service:

  ```bash

>   helm install {{prodnamedash}} projectcalico/tigera-operator --version {{releaseTitle}} -f values.yaml \

Reviewer 2: Have you considered putting all the `--set` flags in a snippet in the way that it can be added to their values.yaml? I suppose this is how most people will be using it.

Perhaps even change it to:
```
echo " <all the options> " >> values.yaml

```

Author: I replaced the list of `--set` with an update to the values
The install command still contains a reference to the pull secret and license `--set-file`

  --namespace tigera-operator --create-namespace
  ```

  or with a LoadBalancer on AWS, add expected annotations to the above installation.

  ```bash
  --set managementCluster.service.annotations[0].key="service.beta.kubernetes.io/aws-load-balancer-type" \
  --set managementCluster.service.annotations[0].value=external \
  --set managementCluster.service.annotations[1].key="service.beta.kubernetes.io/aws-load-balancer-nlb-target-type" \
  --set managementCluster.service.annotations[1].value=instance \
  --set managementCluster.service.annotations[2].key="service.beta.kubernetes.io/aws-load-balancer-scheme" \
  --set managementCluster.service.annotations[2].value=internet-facing \
  ```

  :::note

  The licenseKey must have `ApiVersion: crd.projectcalico.org/v1`.

  :::

1. You can now monitor progress with the following command:

  ```bash
  watch kubectl get tigerastatus
  ```

  Congratulations! You have now installed {{prodname}} for a management cluster using the Helm 3 chart.

</TabItem>
<TabItem label="Managed" value="Managed">

To install a {{prodname}} [managed](../../../multicluster/create-a-managed-cluster#value) cluster with Helm:

> 1. [Configure a storage class for Calico Enterprise](../../../operations/logstorage/create-storage).

Reviewer 2: Managed clusters do not need a storage class.

Author: Removed

  ```bash
  export MANAGEMENT_CLUSTER_ADDR=<your-management-cluster-addr>
  ```

1. Export the management cluster certificate and managed cluster certificate and key.
  
  ```bash
  export MANAGEMENT_CLUSTER_CRT=<management-cluster-crt>
  export MANAGED_CLUSTER_CRT=<managed-cluster-crt>
  export MANAGED_CLUSTER_KEY=<managed-cluster-key>
  ```

1. Install the Tigera {{prodname}} operator and custom resource definitions using the Helm 3 chart:

  ```bash
  helm install {{prodnamedash}} projectcalico/tigera-operator --version {{releaseTitle}} -f values.yaml \
  --set-file imagePullSecrets.tigera-pull-secret=<path/to/pull/secret>,tigera-prometheus-operator.imagePullSecrets.tigera-pull-secret=<path/to/pull/secret> \
  --set-file licenseKeyContent=<path/to/license/file/yaml> \

  --set logStorage.enabled=false --set manager.enabled=false \
  
  --set managementClusterConnection.enabled=true \
  --set managementClusterConnection.address=$MANAGEMENT_CLUSTER_ADDR \
  --set managementCluster.management.tls.crt=$MANAGEMENT_CLUSTER_CRT \
  --set managementCluster.managed.tls.crt=$MANAGED_CLUSTER_CRT \
  --set managementCluster.managed.tls.key=$MANAGED_CLUSTER_KEY \
  

>   --namespace tigera-operator --create-namespace

Reviewer 2: Is this correct?

Author: Does `--create-namespace` work? 

I've successfully tested the behaviour in these docs, and here is a Helm 3 list of install options https://helm.sh/docs/helm/helm_install/#options

Reviewer 2: It must be ok, it just kinda looked funny to me.

  ```

  :::note

>   The licenseKey must have `ApiVersion: crd.projectcalico.org/v1`.

Reviewer 2: Maybe the note can be enriched with a `sed` command to replace the standard ApiVersion. Otherwise this note may leave some question marks to the user.

Author: I added a section to replace the ApiVersion

```
1. Replace the license apiVersion `crd.projectcalico.org/v1`.

  ```bash
  cp <path/to/license/file/yaml> <path/to/licensev1/file/yaml>
  sed -i 's/apiVersion: projectcalico.org\/v3/apiVersion: crd.projectcalico.org\/v1/' <path/to/licensev1/file/yaml>
  ```
```

Author: This has been removed entirely, the license key template file handles the ApiVersion replacement.

### Download the Helm chart

> 1. Add the Calico helm repo:

Reviewer: Delete the numbered bullet. No need for a single item.

       ? `helm install {{prodnamedash}} tigera/tigera-operator --version tigera-operator-v0.0 -f values.yaml \\
--set-file imagePullSecrets.tigera-pull-secret=<path/to/pull/secret>,tigera-prometheus-operator.imagePullSecrets.tigera-pull-secret=<path/to/pull/secret> \\
--set-file licenseKeyContent=<path/to/license/file/yaml> \\
--namespace tigera-operator --create-namespace`
       : `helm install {{prodnamedash}} tigera/tigera-operator --version tigera-operator-{{chart_version_name}} -f values.yaml \\
--set-file imagePullSecrets.tigera-pull-secret=<path/to/pull/secret>,tigera-prometheus-operator.imagePullSecrets.tigera-pull-secret=<path/to/pull/secret> \\
--set-file licenseKeyContent=<path/to/license/file/yaml> \\
--namespace tigera-operator --create-namespace`}
  </CodeBlock>

1. You can now monitor progress with the following command:

  ```bash
  watch kubectl get tigerastatus
  ```

#### Create an admin user and verify management cluster connection.

To access resources in a managed cluster from the {{prodname}} Manager within the management cluster, the logged-in user must have appropriate permissions defined in that managed cluster (clusterrole bindings).

> 1. Create an admin user, `mcm-user`, in the default namespace with full permissions, and token.

Reviewer: Delete numbered bullet.

  ```bash
  kubectl create sa mcm-user
  kubectl create clusterrolebinding mcm-user-admin --clusterrole=tigera-network-admin --serviceaccount=default:mcm-user
  kubectl create token mcm-user -n default
  ```

  Use the generated token, to connect to the UI. In the top right banner in the UI, your management cluster is displayed as the first entry in the cluster selection drop-down menu with the fixed name, `management cluster`.

  Congratulations! You have now installed {{prodname}} for a management cluster using the Helm 3 chart.

</TabItem>
<TabItem label="Managed" value="Managed">

> #### Install a managed cluster using Helm.

Reviewer: Delete. Stem sentence says same thing.

--set-file imagePullSecrets.tigera-pull-secret=<path/to/pull/secret>,tigera-prometheus-operator.imagePullSecrets.tigera-pull-secret=<path/to/pull/secret> \\
--set-file licenseKeyContent=<path/to/license/file/yaml> \\
--set logStorage.enabled=false --set manager.enabled=false \\
--namespace tigera-operator --create-namespace`
       : `helm install {{prodnamedash}} tigera/tigera-operator --version tigera-operator-{{chart_version_name}} -f values.yaml \\
--set-file imagePullSecrets.tigera-pull-secret=<path/to/pull/secret>,tigera-prometheus-operator.imagePullSecrets.tigera-pull-secret=<path/to/pull/secret> \\
--set-file licenseKeyContent=<path/to/license/file/yaml> \\
--set logStorage.enabled=false --set manager.enabled=false \\
--namespace tigera-operator --create-namespace`}
  </CodeBlock>

1. You can now monitor progress with the following command:

  ```bash
  watch kubectl get tigerastatus
  ```

#### Provide permissions to view the managed cluster

  To access resources belonging to a managed cluster from the {{prodname}} Manager UI, the service or user account used to log in must have appropriate permissions defined in the managed cluster.

> 1. Define admin-level permissions for the service account `mcm-user` we created to log in to the Manager UI. Run the following command against your managed cluster.

Reviewer: Delete numbered bullet.

### Download the Helm chart

If you are installing on a cluster installed by EKS, GKE, AKS or Mirantis Kubernetes Engine (MKE), or you need to customize TLS certificates, you **must** customize this Helm chart by creating a `values.yaml` file. Otherwise, you can skip this step.

1. If you are installing on a cluster installed by EKS, GKE, AKS or Mirantis Kubernetes Engine (MKE), set the `kubernetesProvider` as described in the [Installation reference](../../../reference/installation/api.mdx#operator.tigera.io/v1.Provider). For example:

   ```
   echo '{ installation: {kubernetesProvider: EKS }}' > values.yaml
   ```

   For Azure AKS cluster with no Kubernetes CNI pre-installed, create `values.yaml` with the following command:

>    ```

Reviewer: ```suggestion
   ```yaml
```

   cat > values.yaml <<EOF
   installation:
     kubernetesProvider: AKS
     cni:
       type: Calico
     calicoNetwork:
       bgp: Disabled
       ipPools:
       - cidr: 10.244.0.0/16
         encapsulation: VXLAN
   EOF
   ```

> 1. Add any other customizations you require to `values.yaml`. You might like to refer to the [helm docs](https://helm.sh/docs/) or run

Reviewer: ```suggestion
1. Add any other customizations you require to `values.yaml`. To see values that can be customized in the chart, see the [helm docs](https://helm.sh/docs/) or run the following command:
```

>    ```

Reviewer: ```suggestion
   ```bash
```
